src/index.ts

```diff
@@ -0,0 +1,136 @@
+
+const MONTHS_IN_YEAR = 12;
+const MINIMUM_PURCHASE_AMOUNT = 80000;
+
+type DepreciationResult = {
+  year: number;
+  depreciationMonths: number;
+  depreciationAmount: number;
+  percentage: number;
+  startAmount: number;
+  endAmount: number;
+}
+
+type DepreciationInputs = {
+  purchaseAmount: number,
+  purchaseDate: Date,
+  totalDepreciationYears: number
+}
+
+export const assertPurchaseAmount = (amount: number) => {
+  if (isNaN(amount) || amount < 0 || amount % 1 !== 0) { // Check if amount is integer, not float.
+    throw new TypeError('`purchaseAmount` is invalid.');
+  }
+};
+
+export const assertPurchaseDate = (date: Date) => {
+  if (isNaN(date.valueOf())) {
+    throw new TypeError('`purchaseDate` is invalid.');
+  }
+};
+
+export const assertDepreciationYears = (years: number) => {
+  if (isNaN(years) || years < 1 || years % 1 !== 0) { // Check if years is integer, not float.
+    throw new TypeError('`totalDepreciationYears` is invalid.');
+  }
+};
+
+const calculate = (
+  purchaseAmount: number,
+  totalDepreciationYears: number,
+  previousEndAmount: number,
+  monthsLeft: number = MONTHS_IN_YEAR
+): {
+  depreciationAmount: number;
+  percentage: number;
+  startAmount: number;
+  endAmount: number;
+} => {
+  const depreciationAmount = Math.round((purchaseAmount / totalDepreciationYears / MONTHS_IN_YEAR) * monthsLeft);
+  const newEndAmount = previousEndAmount - depreciationAmount;
+
+  // Because of the rounding, even if the calculation is correct, sometimes there is €0.01 left over.
+  // For example, if total is 31, divided by 3 it would be 10 / 10 / 10.
+  if (newEndAmount === 1) {
+    return {
+      depreciationAmount: previousEndAmount,
+      percentage: previousEndAmount / purchaseAmount,
+      startAmount: previousEndAmount,
+      endAmount: 0,
+    }
+  }
+
+  return {
+    depreciationAmount,
+    percentage: depreciationAmount / purchaseAmount,
+    startAmount: previousEndAmount,
+    endAmount: newEndAmount,
+  };
+};
+
+const calculateDepreciation = ({
+  purchaseAmount,
+  purchaseDate,
+  totalDepreciationYears,
+}: DepreciationInputs): DepreciationResult[] => {
+  assertPurchaseAmount(purchaseAmount);
+  assertPurchaseDate(purchaseDate);
+  assertDepreciationYears(totalDepreciationYears);
+
+  const purchaseMonth: number = new Date(purchaseDate).getMonth() + 1;
+  const purchaseYear: number = new Date(purchaseDate).getFullYear();
+  const results: DepreciationResult[] = [];
+
+  let endAmount: number = purchaseAmount;
+  let monthsLeftInLastYear: number = 0;
+
+  // If the price of the good is less than MINIMUM_PURCHASE_AMOUNT, the year is set to 0 automatically.
```

Comment:
```suggestion
  // If the price of the good is less than MINIMUM_PURCHASE_AMOUNT, endAmount is set to 0 automatically.
```

Reply:
`the year is set to 0 automatically` is from one of the AC of [this ticket](https://app.zenhub.com/workspaces/kontist-engineering-5be06ab44b5806bc2bf14eca/issues/kontist/kontax-web/619)

Reply:
I don't get it 🙈 From what I see here you are setting `year` to `purchaseYear` (which is `new Date(purchaseDate).getFullYear();`). Where do you set year to 0?

Reply:
it doesn't matter, the logic for `MINIMUM_PURCHASE_AMOUNT` is removed. :D